src/repr/src/adt/range.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cmp::Ordering;
+use std::error::Error;
+use std::fmt;
+
+use crate::row::DatumNested;
+use crate::Datum;
+
+// Flag constants
+pub const RANGE_EMPTY: u8 = 1;
+pub const RANGE_LB_INC: u8 = 1 << 1;
+pub const RANGE_LB_INF: u8 = 1 << 2;
+pub const RANGE_UB_INC: u8 = 1 << 3;
+pub const RANGE_UB_INF: u8 = 1 << 4;
+
+/// Provides a wrapper to implement the complexities of range semantics over
+/// `RangeInner`.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Range<'a> {
+    /// None value represents empty range
+    pub inner: Option<RangeInner<'a>>,
+}
```

Comment:
This might be a bit more self describing as an enum?

```diff
diff --git a/src/repr/src/adt/range.rs b/src/repr/src/adt/range.rs
index c688bb11e..b0b1bbbfd 100644
--- a/src/repr/src/adt/range.rs
+++ b/src/repr/src/adt/range.rs
@@ -24,9 +24,9 @@ pub const RANGE_UB_INF: u8 = 1 << 4;
 /// Provides a wrapper to implement the complexities of range semantics over
 /// `RangeInner`.
 #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd)]
-pub struct Range<'a> {
-    /// None value represents empty range
-    pub inner: Option<RangeInner<'a>>,
+pub enum Range<'a> {
+    Empty,
+    NonEmpty(NonEmptyRange<'a>),
 }
 
 impl<'a> std::fmt::Display for Range<'a> {
@@ -40,12 +40,12 @@ impl<'a> std::fmt::Display for Range<'a> {
 
 /// Holds the upper and lower `DatumRangeBound`s.
 #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
-pub struct RangeInner<'a> {
+pub struct NonEmptyRange<'a> {
     pub lower: RangeBound<'a>,
     pub upper: RangeBound<'a>,
 }
 ```

This one's just a matter of taste, so take it or leave it!

Reply:
Prefer the ability to `map` on the empty range to preserve `None` values